IOOBE when querying document containing a vector with string "}"

While running property tests (mixed-records-flush-boundary), we observed an error - see failed run [here](https://github.com/xtdb/xtdb/runs/52142348588).

Wrote the following minimal failing test against this:
```
(t/deftest string-ioobe-repro
  (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 1 :a ["}"]}]]) 
  (t/is (= [{:xt/id 1 :a ["}"]}] (xt/q tu/*node* "SELECT * FROM docs ORDER BY _id"))))
```

Error stack trace:
```
Caused by: java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.remove(ArrayList.java:551)
	at org.postgresql.jdbc.ArrayDecoding.buildArrayList(ArrayDecoding.java:691)
	at org.postgresql.jdbc.PgArray.buildArrayList(PgArray.java:309)
	at org.postgresql.jdbc.PgArray.getArrayImpl(PgArray.java:165)
	at org.postgresql.jdbc.PgArray.getArray(PgArray.java:116)
	at xtdb.next.jdbc$read_col.invokeStatic(jdbc.clj:15)
	at xtdb.next.jdbc$read_col.invokePrim(jdbc.clj)
	at xtdb.next.jdbc$builder_fn$reify__3970.with_column(jdbc.clj:43)
	at next.jdbc.result_set$row_builder$fn__1648.invoke(result_set.clj:447)
	at clojure.lang.LongRange.reduce(LongRange.java:170)
	at clojure.core$reduce.invokeStatic(core.clj:6964)
	at clojure.core$reduce.invoke(core.clj:6947)
	at next.jdbc.result_set$row_builder.invokeStatic(result_set.clj:447)
	at next.jdbc.result_set$row_builder.invoke(result_set.clj:443)
	at next.jdbc.result_set$mapify_result_set$reify__1709.iterator(result_set.clj:519)
	at clojure.core.protocols$iter_reduce.invokeStatic(protocols.clj:52)
	at clojure.core.protocols$fn__8260.invokeStatic(protocols.clj:74)
	at clojure.core.protocols$fn__8260.invoke(protocols.clj:74)
	at clojure.core.protocols$fn__8207$G__8202__8220.invoke(protocols.clj:13)
	at clojure.core$reduce.invokeStatic(core.clj:6965)
	at clojure.core$into.invokeStatic(core.clj:7038)
	at clojure.core$into.invoke(core.clj:7029)
	at xtdb.api$plan_q$reify__4270$fn__4271$fn__4272.invoke(api.clj:124)
	at clojure.core$map$fn__5952$fn__5953.invoke(core.clj:2759)
	at next.jdbc.result_set$reduce_result_set.invokeStatic(result_set.clj:709)
	at next.jdbc.result_set$reduce_result_set.invoke(result_set.clj:701)
	at next.jdbc.result_set$reduce_stmt.invokeStatic(result_set.clj:725)
	at next.jdbc.result_set$reduce_stmt.invoke(result_set.clj:715)
	at next.jdbc.result_set$fn$reify__1772.reduce(result_set.clj:872)
	at clojure.core$transduce.invokeStatic(core.clj:7025)
	at clojure.core$transduce.invoke(core.clj:7012)
	at xtdb.api$plan_q$reify__4270$fn__4271.invoke(api.clj:124)
```

--- a/core/src/main/clojure/xtdb/pgwire/types.clj
+++ b/core/src/main/clojure/xtdb/pgwire/types.clj
@@ -99,6 +99,20 @@
 
 (defn- read-utf8 [^bytes barr] (String. barr StandardCharsets/UTF_8))
 
+(defn- escape-pg-array-element
+  "Escapes a string element for PostgreSQL array format.
+  Wraps in quotes if it contains special characters or would otherwise be misparsed."
+  [^String s]
+  (let [needs-quotes? (or (str/blank? s)
+                          (#{"NULL"} s)
+                          (re-find #"[\"\\,\{\}\s]" s))]
+    (if needs-quotes?
+      (format "\"%s\""
+              (-> s
+                  (str/replace "\\" "\\\\")
+                  (str/replace "\"" "\\\"")))
+      s)))
+
 (defn utf8
   "Returns the utf8 byte-array for the given string"
   ^bytes [s]
@@ -850,7 +864,9 @@
                                (let [list (.getObject list-rdr idx)
                                      sb (StringBuilder. "{")]
                                  (doseq [elem list]
-                                   (.append sb (or elem "NULL"))
+                                   (if (nil? elem)
+                                     (.append sb "NULL")
+                                     (.append sb (escape-pg-array-element (str elem))))
                                    (.append sb ","))
                                  (if (seq list)
                                    (.setCharAt sb (dec (.length sb)) \})

--- a/core/src/main/clojure/xtdb/pgwire/types.clj
+++ b/core/src/main/clojure/xtdb/pgwire/types.clj
@@ -99,19 +99,11 @@
 
 (defn- read-utf8 [^bytes barr] (String. barr StandardCharsets/UTF_8))
 
-(defn- escape-pg-array-element
-  "Escapes a string element for PostgreSQL array format.
-  Wraps in quotes if it contains special characters or would otherwise be misparsed."
-  [^String s]
-  (let [needs-quotes? (or (str/blank? s)
-                          (#{"NULL"} s)
-                          (re-find #"[\"\\,\{\}\s]" s))]
-    (if needs-quotes?
-      (format "\"%s\""
-              (-> s
-                  (str/replace "\\" "\\\\")
-                  (str/replace "\"" "\\\"")))
-      s)))
+(defn- escape-pg-array-element [^String s]
+  (format "\"%s\""
+          (-> s
+              (str/replace "\\" "\\\\")
+              (str/replace "\"" "\\\""))))
 
 (defn utf8
   "Returns the utf8 byte-array for the given string"

--- a/src/test/clojure/xtdb/node_test.clj
+++ b/src/test/clojure/xtdb/node_test.clj
@@ -1219,3 +1219,49 @@ VALUES(1, OBJECT (foo: OBJECT(bibble: true), bar: OBJECT(baz: 1001)))"]])
             {:column-name "_valid_to"}
             {:column-name "h"}]
            (xt/q tu/*node* "SELECT column_name FROM information_schema.columns WHERE table_name = 'docs'"))))
+
+(t/deftest test-string-escaping-in-vectors
+  (t/testing "curly braces - #4858"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 1 :a ["}"]}]])
+    (t/is (= [{:xt/id 1 :a ["}"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 1"))))
+
+  (t/testing "question mark and curly brace - #4859"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 2 :a ["?{"]}]])
+    (t/is (= [{:xt/id 2 :a ["?{"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 2"))))
+
+  (t/testing "commas"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 3 :a ["," ",hello" "hello,world"]}]])
+    (t/is (= [{:xt/id 3 :a ["," ",hello" "hello,world"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 3"))))
+
+  (t/testing "spaces"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 4 :a ["a " "b " " c " "hello world"]}]])
+    (t/is (= [{:xt/id 4 :a ["a " "b " " c " "hello world"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 4"))))
+
+  (t/testing "quotes"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 5 :a ["\"quoted\""]}]])
+    (t/is (= [{:xt/id 5 :a ["\"quoted\""]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 5"))))
+
+  (t/testing "backslashes"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 6 :a ["a\\b\\c"]}]])
+    (t/is (= [{:xt/id 6 :a ["a\\b\\c"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 6"))))
+
+  (t/testing "empty string"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 7 :a [""]}]])
+    (t/is (= [{:xt/id 7 :a [""]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 7"))))
+
+  (t/testing "literal NULL string"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 8 :a ["NULL"]}]])
+    (t/is (= [{:xt/id 8 :a ["NULL"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 8"))))
+  
+  (t/testing "escape characters"
+    (xt/execute-tx tu/*node* [[:put-docs :docs {:xt/id 9 :a ["\n" "\t" "\r"]}]])
+    (t/is (= [{:xt/id 9 :a ["\n" "\t" "\r"]}]
+             (xt/q tu/*node* "SELECT * FROM docs WHERE _id = 9")))))

